### ISSUE TITLE
Handle user name purification outliers

### ISSUE THREAD

**Issue by DanilYachmenev** [sorrentum/sorrentum#446]
There is a function ‎`purify_from_environment()` in `helpers/hunit_test.py` that replaces the path to user names with `$USER_NAME`.
E.g. `root` is replaced with `$USER_NAME` since `root` is a username.

However there are some false positives, e.g.,
`take_square_root` -> `take_square_$USER_NAME`
This is handled manually by checking if `take_square_root` is in the text. However, the goal is to come up with a more general approach.
`out_col_groups: [('root_q_mv',), ('root_q_mv_adj',), ('root_q_mv_os',)]` -> `out_col_groups: [('$USER_NAME_q_mv',), ('$USER_NAME_q_mv_adj',), ('$USER_NAME_q_mvs_os’,)]`

The goal to:
- [x] Modify the text replacement such that the function replaces `root` with `$USER_NAME` only when `root` is a user. Maybe we can just say that if `root` is a stand-alone word it is a username, otherwise it’s not.
- [x] Add a unit test to https://github.com/sorrentum/sorrentum/blob/master/helpers/test/test_unit_test.py (use the tests in the class as a reference)
Create a new test class `Test_purify_from_environment`
Use `take_square_root` and `[('root_q_mv',), ('root_q_mv_adj',), ('root_q_mv_os',)]` as inputs
For more info about unit tests see the corresponding [doc](https://github.com/sorrentum/sorrentum/blob/master/docs/Unit_tests.md)

FYI @PomazkinG

**Comment by samarth9008** [comment]
Yeah no error on my side as well. Can you run `helpers/test/test_unit_test.py::TestCheckDataFrame1::test_check_df_equal1 -s --dbg` and post the relevant logs? This will give us more detailed reason for failing.

**Comment by PomazkinG** [comment]
@dreamcatcher0101 another question: did you run the tests from a Docker container?

In theory if you run from a Docker container then the env should be properly tuned

**Comment by dreamcatcher0101** [comment]
I ran `pytest helpers/test/test_unit_test.py::TestCheckDataFrame1::test_check_df_equal1 -s --dbg`.
Result is
```
python: 3.8.10
...
pandas: 2.0.3
```

**Comment by PomazkinG** [comment]
the pandas is version is wrong. @dreamcatcher0101 can you try to start a Docker container before running the regressions?
```
> i docker_bash
> pytest helpers/test/test_unit_test.py
```

docs: https://github.com/sorrentum/sorrentum/blob/master/docs/Quick_start_for_developing.md#sorrentum-dev-docker-container-aka-dev-container-cmamp-container

**Comment by dreamcatcher0101** [comment]
I tried to start Docker container and was able to fix the error.
Is Pandas version `1.5.1`?
But now another error occurs. 
```
Traceback (most recent call last):
  File "/app/helpers/test/test_unit_test.py", line 1202, in test1
    self.helper(txt, exp)
  File "/app/helpers/test/test_unit_test.py", line 1192, in helper
    act = hunitest.purify_from_environment(txt)
  File "/app/helpers/hunit_test.py", line 429, in purify_from_environment
    user_name = hsystem.get_user_name()
  File "/app/helpers/hsystem.py", line 73, in get_user_name
    res = getpass.getuser()
  File "/usr/lib/python3.8/getpass.py", line 169, in getuser
    return pwd.getpwuid(os.getuid())[0]
KeyError: 'getpwuid(): uid not found: 1000'
```

**Comment by PomazkinG** [comment]
@dreamcatcher0101 the problem is that there is no user with id `1000` inside a Docker container

1) What is your OS?  
2) Run this outside a Docker container and please post the output:
```
# Should work for Linux and Mac.
> id -u
```

**Comment by dreamcatcher0101** [comment]
@PomazkinG Thanks for your response.
My OS is Ubuntu.
I run command `i docker_bash` and here's result:
![image](https://github.com/sorrentum/sorrentum/assets/102745567/93dc09bb-1902-4201-931b-966e7a23f49a)

**Comment by samarth9008** [comment]
Run this outside the docker container

**Comment by dreamcatcher0101** [comment]
@samarth9008 
Here's the result.
![image](https://github.com/sorrentum/sorrentum/assets/102745567/4c09f067-c45b-40f6-a1dc-d77deb9a5180)

**Comment by PomazkinG** [comment]
the problem is on our side then, let me consult with @gpsaggese 

From my POV the issue is that we create users starting from `id=1001` while people might have `id=1000` (which should be the first id available for a regular user):

> Reserved UID ranges: Linux reserves certain ranges of UIDs for specific purposes. For example, UIDs 0 to 999 are usually reserved for system users, and UIDs 1000 and above are typically used for regular user accounts. 

https://github.com/sorrentum/sorrentum/blob/8f557a920251a972b1f6e3190141f03eee226048/devops/docker_build/create_users.sh#L16-L20

The options are:

1) Re-release the cmamp / sorrentum image with user id starting point = 1000. The question here: was there any specific reason to start with 1001?
2) Ask people to change the UID to smth that fits into our range -- could be dangerous since once to carefully change the ownership of the files
3) Allow people to run `i docker_bash` as `root` -- could potentially introduce permission issues

**Comment by dreamcatcher0101** [comment]
Hi @PomazkinG.
Would you assign me another issue so that I can keep going on working while you are consulting?

**Comment by gpsaggese** [comment]
I've noticed the same error and proposed the same solutions (https://github.com/sorrentum/sorrentum/issues/444#issuecomment-1652513842)

1) I don't think there is any reason for starting from 1001. So I would fix in the next container release
2) as a bandaid I would actually fix up the container doing a `useradd` + `usermod` in the container and then push as dev. @dreamcatcher0101 you could fix up your local copy of the Docker image, if you are handy with Docker (of course you can't push the container).
3) as horrible hack to keep going, we could remap the 1000
```
val = os.getuid()
if val == 1000:
    val = 1001
...
...
```
The problem is that the consequences when writing files from Docker would be unclear, the perms might be screwed up and who knows what else...

Personally I would 2) to unblock the Ubuntu users, fix up `sorrentum/devops/docker_build/create_users.sh` and then when we do the next release we get it fixed properly.

Good job debugging / finding the solution.

**Comment by PomazkinG** [comment]
@dreamcatcher0101 just mentioning another option to keep making progress https://github.com/sorrentum/sorrentum/issues/444#issuecomment-1652740499

**Comment by PomazkinG** [comment]
@dreamcatcher0101 please address the changes in https://github.com/sorrentum/sorrentum/pull/465

**Comment by DanilYachmenev** [comment]
done
